### ISSUE TITLE
DNS udp_answer_limit is not applied if bufsize != 512

### ISSUE THREAD

**Issue by pierresouchay** [hashicorp/consul#3937]
When the DNS configuration includes a `udp_answer_limit`, this parameter is ONLY applied if the buffer size is set to 512 bytes.

Reproducing it:

Set `udp_answer_limit = 1` in consul configuration

```
$ dig @localhost +bufsize=512 SRV myservice.service.consul -p 8600 +short
1 1 80 host-001.acme.com.
```

```
$ dig @localhost +bufsize=513 SRV myservice.service.consul -p 8600 +short
1 1 80 host-001.acme.com.
1 1 80 host-002.acme.com.
```

It is a big deal for us since TXT records have been added since Consul 1.x => thus the SRV answers are far more bigger since it include node metadata encoded as TXT records.

Thus, the 512 bytes limits is likely to be reached (it is for us) => SRV records return no responses at all with Consul > 1.x since the size now overflows 512 bytes

**Comment by pierresouchay** [comment]
Same issue as https://github.com/hashicorp/consul/issues/3355

**Comment by pierresouchay** [comment]
OK, I figured out what is the problem: the configuration `udp_answer_limit` has wrong semantics.

When you look at the source code, there is this paragraph https://github.com/hashicorp/consul/blob/master/agent/config/runtime.go#L255 : 

```
        // DNSUDPAnswerLimit is used to limit the maximum number of DNS Resource
	// Records returned in the ANSWER section of a DNS response. This is
	// not normally useful and will be limited based on the querying
	// protocol, however systems that implemented §6 Rule 9 in RFC3484
	// may want to set this to `1` in order to subvert §6 Rule 9 and
	// re-obtain the effect of randomized resource records (i.e. each
	// answer contains only one IP, but the IP changes every request).
	// RFC3484 sorts answers in a deterministic order, which defeats the
	// purpose of randomized DNS responses.  This RFC has been obsoleted
	// by RFC6724 and restores the desired behavior of randomized
	// responses, however a large number of Linux hosts using glibc(3)
	// implemented §6 Rule 9 and may need this option (e.g. CentOS 5-6,
        // Debian Squeeze, etc).
```
However, in the code, in https://github.com/hashicorp/consul/blob/master/agent/dns.go#L742 :
```
        // This cuts UDP responses to a useful but limited number of responses.
	maxAnswers := lib.MinInt(maxUDPAnswerLimit, udpAnswerLimit)
        if maxSize == defaultMaxUDPSize && numAnswers > maxAnswers {
```

This code basically check whether the size of UDP is different from 512, which is DNS without EDNS.

Since `maxUDPAnswerLimit` is hardcoded to 8, it simply allow to remove a few entries when EDNS is not used.

So, if you want to use the mechanism to have a single value returned each time, this will work ONLY with non EDNS records => which means the first comment in https://github.com/hashicorp/consul/blob/master/agent/config/runtime.go#L255 is wrong => it cannot be used to have randomized records => a problem for us.

The easy patch would be to change the test https://github.com/hashicorp/consul/blob/master/agent/dns.go#L742 to something like:

```
  if numAnswers > maxAnswers {
```

exactly like it was done in https://github.com/hashicorp/consul/pull/3356

However, this change has a few issues:
* it would change the behavior of Consul
* it also applies to SRV records (which do not suffer from the randomization issue - since we want to get all the records)

Thus, I propose the following change:
* keep, but depreciate in documentation `udp_answer_limit` => this parameter is actually working for UDP without EDNS, but has not the effect of disabling randomization, thus, it is a poor optimization (since it allow to reduce from 8 to 3 records in requests for instance)
* add a new configuration option in `dns_config` called `a_record_limit` that would return only `a_record_limit` items for A, AAAA, ANY queries, but would not apply to SRV queries

What do you think of this change ?

I am preparing a PR to enable this behavior

**Comment by pierresouchay** [comment]
Note that this patch really address randomization issue as it also works with DNS TCP connections as well as UDP/EDNS and UDP DNS 512 bytes.

It is also avoid breaking SRV discovery due to randomization workarounds on DNS resolvers

**Comment by pierresouchay** [comment]
Patch has been merged for consul >= 1.0.7
